fix jl_active_task_stack to return correct values

vtjnash opened the pull request:
Store the activefp at every point where it might be observed to have changed. Seems potentially slightly costly to need to generate these extra writes for every safepoint transition to keep track of the current stack location, but that is potentially an unavoidable cost to doing conservative stack scanning for gc.

Replace #54923

vtjnash committed:
fix jl_active_task_stack to return correct values

Seems potentially slightly costly to need to generate these extra
writes for every safepoint transition to keep track of the current
stack location, but that is potentially an unavoidable cost to doing
conservative stack scanning for gc.

fingolfin commented:
This would be *really* nice to have :-)

fingolfin commented on src/julia_threads.h:
```diff
@@ -214,6 +215,68 @@ typedef struct _jl_tls_states_t {
+typedef struct _jl_excstack_t jl_excstack_t;
+typedef struct _jl_handler_t jl_handler_t;
+
+typedef struct _jl_task_t {
```
A major chunk of this PR seems to be about moving `_jl_task_t` from `julia.h` to here. Is there a functional reason why this is necessary for this PR? Perhaps something got added to `_jl_task_t` -- but it's difficult to say because everything is in a single commit

fingolfin replied:
Made PR #57117 to take care of this

fingolfin commented on src/gc-stock.c:
```diff
@@ -3335,6 +3335,7 @@ JL_DLLEXPORT void jl_gc_collect(jl_gc_collection_t collection)
     }
     jl_gc_debug_print();
 
+    ct->ctx.activefp = (char*)jl_get_frame_addr();
```
OK, that one seems simple enough :-)

fingolfin commented on src/julia_internal.h:
```diff
@@ -1720,18 +1720,6 @@ JL_DLLEXPORT unsigned jl_special_vector_alignment(size_t nfields, jl_value_t *fi
 void register_eh_frames(uint8_t *Addr, size_t Size) JL_NOTSAFEPOINT;
 void deregister_eh_frames(uint8_t *Addr, size_t Size) JL_NOTSAFEPOINT;
 
-STATIC_INLINE void *jl_get_frame_addr(void) JL_NOTSAFEPOINT
```
Moved to `julia_threads.h` so it can be called there, OK

fingolfin commented on src/julia_threads.h:
```diff
@@ -270,6 +345,9 @@ STATIC_INLINE int8_t jl_gc_state_set(jl_ptls_t ptls, int8_t state,
     assert(old_state != JL_GC_CONCURRENT_COLLECTOR_THREAD);
+    if (__builtin_constant_p(state) ? state : // required if !old_state && state, otherwise optional
+        __builtin_constant_p(old_state) ? !old_state : 1)
+        jl_atomic_load_relaxed(&ptls->current_task)->ctx.activefp = (char*)jl_get_frame_addr();
```
I have no idea what the semantics for `jl_gc_state_set` are, when it is supposed to be called, etc., so I can't judge whether this is sensible or effective.

This seems to store the current frame pointer into the task. Presumably that means this function is being called from the task as it is about to switch out? At least that would make sense.

My personal dream is that the wonderful people adding MMTk and doing other GC restructuring work, like @gbaraldi @d-netto @udesou), add some more comments describing such things as they figure this stuff out for themselves ;-).

d-netto replied:
Not sure if we should be updating the `activefp` in `jl_gc_state_set`?

Makes more sense to me to update this in the `segv_handler` or `jl_set_gc_and_wait`.

Also, I believe we should fetch the frame pointer from the context the thread was running with before it got trapped in the signal handler -- note that we run the signal handlers with a different stack.

This is more or less what I started doing in https://github.com/JuliaLang/julia/pull/55003.

fingolfin replied:
Well, we already do use conservative GC in Julia 1.11 and master... though we do it by catching the "segfaults", though really the only ones we had so far were due to incorrect values for the stack *start* being reported causing us to read into guard pages.

Though perhaps we'd be crashing more if we used tasks extensively? I don't know enough about what the problem is right now / how "wrong" the currently reported values are? I thought they are just "too big" as in the stack is overestimated? But maybe that's not true anymore?

fingolfin replied:
I still don't quite understand what that means. Anyway based on the bits I do understand I hope we won't be affected because we don't scan for general Julia objects, just for those of a specific "foreign" type. Since these are allocated via the C ABI from C code I don't think the LLVM optimizer has any chance of optimizing them away in any way...

fingolfin commented on src/llvm-codegen-shared.h:
```diff
@@ -256,16 +256,24 @@ static inline void emit_gc_safepoint(llvm::IRBuilder<> &builder, llvm::Type *T_s
+        Value *currentfp = builder.CreateConstInBoundsGEP1_32(T_int8, task, offset, "gc_state");
+        Value *fp = builder.CreateIntrinsic(Intrinsic::frameaddress, {builder.getPtrTy()}, {builder.getInt32(0)});
+        builder.CreateAlignedStore(fp, currentfp, Align(sizeof(void*)));
+    }
```
I have no idea how this code is supposed to work, so I can't review it other than on a superficial level: My guess would be that this function is a code-emitting counterpart to `jl_gc_state_set`, and so a similar modification as there is applied here, to update `activefp` to contain the current framepointer. The inserted code looks plausibly like it might be doing that

fingolfin commented on src/task.c:
```diff
@@ -345,81 +345,39 @@ void JL_NORETURN jl_finish_task(jl_task_t *ct)
-    return (void *)((char *)task->ctx.stkbuf + off);
-}
-
 JL_DLLEXPORT void jl_active_task_stack(jl_task_t *task,
```
THe changes in `jl_active_task_stack` look plausible to me

fingolfin reviewed (approved):
Looks good to me to the extent that I can evaluate it. I really hope people like @JeffBezanson @d-netto @gbaraldi @udesou might find time to have a look so it can be merged

fingolfin committed:
Remove jl_task_stack_buffer

This was removed previously in PR #54527 but had to be reverted in
PR #54559 as one usage remained (more by accident then by design).
This has since then been resolved.

fingolfin committed:
Move jl_task_t to julia_threads.h
